File: worker/src/greeting.ts
/** Основная логика: на /start и на слова-триггеры — мем плюс привет. */
import { Composer, GrammyError } from "grammy";
import type { BotContext } from "./types";
import { sendMedia } from "./media";
import { matchesTrigger } from "./triggers";

// Сколько битых file_id подряд готовы пережить, прежде чем ответить без мема.
const MAX_MEME_ATTEMPTS = 5;

export function escapeHtml(s: string): string {
  return s
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;");
}

/** Подставляет {name} и {mention} в шаблон приветствия. */
export function renderGreeting(
  template: string,
  userId: number,
  firstName: string,
): string {
  const name = escapeHtml(firstName || "друг");
  const mention = `<a href="tg://user?id=${userId}">${name}</a>`;
  return template.split("{name}").join(name).split("{mention}").join(mention);
}

/** Шлёт случайный мем. Битые file_id помечает и пробует следующий. */
async function sendRandomMeme(ctx: BotContext): Promise<number | null> {
  const chatId = ctx.chat?.id;
  if (chatId === undefined) return null;

  for (let attempt = 0; attempt < MAX_MEME_ATTEMPTS; attempt++) {
    const meme = await ctx.db.pickMeme(chatId);
    if (!meme) return null;
    try {
      await sendMedia(ctx, meme.file_id, meme.kind, meme.caption);
    } catch (err) {
      if (!(err instanceof GrammyError)) throw err;
      console.warn(`Мем #${meme.id} не отправился (${err.description}), помечаю битым`);
      await ctx.db.markBroken(meme.id);
      continue;
    }
    await ctx.db.setLastMeme(chatId, meme.id);
    return meme.id;
  }
  return null;
}

export async function greet(ctx: BotContext): Promise<void> {
  const user = ctx.from;
  const chatId = ctx.chat?.id;
  if (!user || user.is_bot || chatId === undefined) return;

  if (!(await ctx.db.isEnabled())) return;

  const cooldown = await ctx.db.cooldownMinutes();
  if (!(await ctx.db.canGreet(chatId, user.id, cooldown))) return;

  const memeId = await sendRandomMeme(ctx);

  const text = renderGreeting(await ctx.db.greetingText(), user.id, user.first_name);
  try {
    await ctx.reply(text, { parse_mode: "HTML" });
  } catch (err) {
    if (!(err instanceof GrammyError)) throw err;
    // Кривая разметка в шаблоне — отправляем как есть, без HTML.
    console.warn("Приветствие не разобралось как HTML, шлю текстом");
    await ctx.reply(text);
  }

  await ctx.db.logGreet(chatId, user.id, memeId);

  if (memeId === null && ctx.adminIds.has(user.id) && (await ctx.db.countMemes()) === 0) {
    await ctx.reply(
      "☝️ Пул мемов пуст. Кинь мне картинку с подписью <code>/addmeme</code> " +
        "или ответь на неё этой командой.",
      { parse_mode: "HTML" },
    );
  }
}

export const greetingComposer = new Composer<BotContext>();

greetingComposer.command("start", greet);

greetingComposer.on("message", async (ctx, next) => {
  const text = ctx.message.text ?? ctx.message.caption ?? "";
  if (!text || text.startsWith("/")) return next();
  if (matchesTrigger(text, await ctx.db.listTriggers())) {
    await greet(ctx);
    return;
  }
  return next();
});

File: bot/handlers/greeting.py
"""Основная логика: на /start и на слова-триггеры — случайный мем плюс привет."""
from __future__ import annotations

import html
import logging

from aiogram import Router
from aiogram.exceptions import TelegramBadRequest
from aiogram.filters import CommandStart
from aiogram.types import Message, User

from bot.db import Database
from bot.filters import HasTrigger
from bot.media import send_media

logger = logging.getLogger(__name__)
router = Router(name="greeting")

# Сколько битых file_id подряд готовы пережить, прежде чем ответить без мема.
MAX_MEME_ATTEMPTS = 5


def render_greeting(template: str, user: User) -> str:
    """Подставляет {name} и {mention}.

    Через replace, а не format: в шаблоне админа могут быть свои фигурные
    скобки, и format на них упал бы.
    """
    name = user.first_name or "друг"
    return template.replace("{name}", html.escape(name)).replace(
        "{mention}", user.mention_html(name)
    )


async def _send_random_meme(message: Message, db: Database) -> int | None:
    """Шлёт случайный мем. Битые file_id помечает и пробует следующий."""
    for _ in range(MAX_MEME_ATTEMPTS):
        meme = await db.pick_meme(message.chat.id)
        if meme is None:
            return None
        try:
            await send_media(message, meme["file_id"], meme["kind"], meme["caption"])
        except TelegramBadRequest as exc:
            logger.warning("Мем #%s не отправился (%s), помечаю битым", meme["id"], exc)
            await db.mark_broken(meme["id"])
            continue
        await db.set_last_meme(message.chat.id, meme["id"])
        return meme["id"]
    return None


async def greet(message: Message, db: Database, admin_ids: frozenset[int]) -> None:
    user = message.from_user
    if user is None or user.is_bot:
        return

    if not await db.is_enabled():
        logger.debug("Бот выключен, молчу")
        return

    cooldown = await db.cooldown_minutes()
    if not await db.can_greet(message.chat.id, user.id, cooldown):
        logger.debug("Кулдаун для %s в чате %s", user.id, message.chat.id)
        return

    meme_id = await _send_random_meme(message, db)

    text = render_greeting(await db.greeting_text(), user)
    try:
        await message.answer(text)
    except TelegramBadRequest:
        # Кривая разметка в шаблоне — отправляем как есть, без HTML.
        logger.warning("Приветствие не разобралось как HTML, шлю текстом")
        await message.answer(text, parse_mode=None)

    await db.log_greet(message.chat.id, user.id, meme_id)

    if meme_id is None and user.id in admin_ids and await db.count_memes() == 0:
        await message.answer(
            "☝️ Пул мемов пуст. Кинь мне картинку с подписью <code>/addmeme</code> "
            "или ответь на неё этой командой."
        )


@router.message(CommandStart())
async def on_start(message: Message, db: Database, admin_ids: frozenset[int]) -> None:
    await greet(message, db, admin_ids)


@router.message(HasTrigger())
async def on_trigger(message: Message, db: Database, admin_ids: frozenset[int]) -> None:
    await greet(message, db, admin_ids)
